### PR TITLE
Filter tool: fix too strict sanitizing

### PR DESCRIPTION
**Pull request by abretaud**
Hi,
There is problem with the Filter1 tool: the given example does not work as there is a little too much sanitization of the user input.
ie `c1=='chr22'` gets sanitized as `c1==__sq__chr22__sq__`
This is also a problem when matching on strings more complicated than chr22 (like 100.00% -> the % becomes a X)
Here's a patch to fix it by allowing most characters, just escaping $ to avoid getting injecting bash variables

### CODE REVIEW

**Comment by bgruening** [comment]
Is `"` working?

File: tools/stats/filtering.xml [hunk 1]
```diff
@@ -10,6 +10,14 @@
     <param format="tabular" name="input" type="data" label="Filter" help="Dataset missing? See TIP below."/>
     <param name="cond" size="40" type="text" value="c1=='chr22'" label="With following condition" help="Double equal signs, ==, must be used as shown above. To filter for an arbitrary string, use the Select tool.">
       <validator type="empty_field" message="Enter a valid filtering condition, see syntax and examples below."/>
+      <sanitizer>
+        <valid initial="string.printable">
```
Comment (hexylena) [inline tools/stats/filtering.xml]: string.printable will include `"` which will affect the string around cond. Maybe we could use `'$cond'` in the command, and then string.printable here without `'`?

**Comment by abretaud** [comment]
yeah, probably better like this
will the tool tests be run for tools in the galaxy code?

**Comment by nsoranzo** [comment]
> will the tool tests be run for tools in the galaxy code?

I don't think they are run, but you can do it locally with Planemo.

**Comment by jmchilton** [comment]
This is really opening things wide up, I'm not sure how to hack it - I'm not sure it is hackable - but it makes me nervous. Are others confident that nothing bad can get through here and unquote that string? I'm not -1 since I don't know how to break it though.

An alternative worth considering is reading ``cond`` from a JSON file created by Galaxy with:

```xml
    <configfiles>
        <inputs name="inputs" />
    </configfiles>
```

An example is here (https://github.com/galaxyproject/galaxy/blob/dev/test/functional/tools/inputs_as_json.xml). The command might then be:

```
filtering.py $input $out_file1 '$inputs' ${input.metadata.columns} "${input.metadata.column_types}" $header_lines
```

And then in filtering you might be able to do something like:

```python
import json
open(sys.argv[3], "r") as f:
    inputs = json.load(f)
cond_text = inputs["cond"]
```

**Comment by abretaud** [comment]
Done, it looks safer indeed!
I just realized that there was some unsanitizing done at https://github.com/abretaud/galaxy/blob/0b191abaf18271b9a51f21a13808e2b74c6cf0b1/tools/stats/filtering.py#L171. But it could not restore the ambiguous X sign (for % for example)

**Comment by jmchilton** [comment]
[View Tool Test Results](http://galaxyproject.github.io/planemo/tool_test_viewer.html?test_data_url=https://gist.githubusercontent.com/jmchilton/19cead60138cd36e45e985dfb635bcee/raw/2f4b9b891bfea2ba4d832d068de025c49966b1e8/index) - they look good to me.

We don't change the interface here in a way that would break or change an existing run - we just let more expressions through than previously - so I think we should not change the version. (Even if we would change it in the tool shed - it is harder to deal with multiple versions of bundled tools still so I'd prefer to not change the version here.)

If you make that change, I will merge this - looks great to me!

**Comment by jmchilton** [comment]
Well I didn't read your last comment closely enough - do you think this isn't quite ready to go yet?

**Comment by abretaud** [comment]
I think it's ready, it's just that the bug fixed here was not so common than I thought: the tool failed only when there was some char like % or &, but not with quotes.
I've added a test case that fails without the patch

**Comment by jmchilton** [comment]
Great - thanks a bunch @abretaud !

**Pull request merged**